### PR TITLE
expression: Fix unexpected panic when using IF function. (#21132)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #21132 to release-4.0

---

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/15847 

Problem Summary:
When rewrite expression, IF(null column,null,null) will get an evalTp TpString. And it will wrap arg0 with isTruth.
Above IF function, there has a CastToReal function.   
So the final expression is     CastStringToReal(IFString(isTruth(null column), null , null))
In other optimization, maybe it will refactor the expression.
IFString(isTruth(null column), null, null)  will be refactored to IFINT(isTruth(null column), null, null)
(See InferType4ControlFuncs and typeInferForNull)
It will cause panic, since CastStringToReal will try to call evalString().

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:

### Related changes

- Need to cherry-pick to the release branch 4.0

### Check List 

Tests 

- Unit test
- Integration test

Side effects

### Release note 

- Fix unexpected panic when using IF function

### CODE REVIEW

File: expression/integration_test.go [hunk 1]
```diff
@@ -7282,6 +7282,75 @@ func (s *testIntegrationSuite) TestIssue20860(c *C) {
 	c.Assert(tk.ExecToErr("update t set d = adddate(d, interval 1 day) where id < 10"), NotNil)
 }
 
+<<<<<<< HEAD
```
Comment (XuHuaiyu) [inline expression/integration_test.go:7285]: conflicts

**Comment by wshwsh12** [comment]
@lzmhhh123 @XuHuaiyu  PTAL

**Review by qw4990 — COMMENTED** [review]
LGTM

**Comment by qw4990** [comment]
@wshwsh12 Please resolve conflicts again...

**Review by zz-jason — APPROVED** [review]
LGTM

**Pull request merged**